test/uk/gov/hmrc/tai/controllers/income/IncomeControllerSpec.scala

```diff
@@ -191,15 +200,452 @@ class IncomeControllerSpec extends PlaySpec
+        )))
+
+      val mockEmploymentService = mock[EmploymentService]
+      when(mockEmploymentService.employments(any[Nino], any[TaxYear])(any[HeaderCarrier]))
+        .thenReturn(Future.successful(employments))
+
+      val sut = createSUT(incomeService = mockIncomeService, employmentService = mockEmploymentService, authentication = loggedInAuthenticationPredicate)
+      val result = sut.matchedTaxCodeIncomesForYear(nino, TaxYear().next, "EmploymentIncome", "PotentiallyCeased")(FakeRequest())
+
+      status(result) mustBe OK
+
+      val expectedIncomeSource = Json.toJson(IncomeSource(TaxCodeIncome(EmploymentIncome, Some(2), BigDecimal(0),
+        "EmploymentIncome", "1100L", "Employer2", OtherBasisOperation, PotentiallyCeased, BigDecimal(321.12), BigDecimal(0), BigDecimal(0)),
+        employment
+      ))
+
+      val expectedJson = Json.obj(
+        "data" -> Json.arr(expectedIncomeSource),
+        "links" -> Json.arr()
+      )
+
+      contentAsJson(result) mustBe expectedJson
+    }
+
+    "return empty JSON when no records match" in {
+      val mockIncomeService = mock[IncomeService]
+      when(mockIncomeService.taxCodeIncomes(any(), Matchers.eq(TaxYear().next))(any()))
+        .thenReturn(Future.successful(Seq(
+          TaxCodeIncome(EmploymentIncome, None, BigDecimal(0),
+            "EmploymentIncome", "1100L", "Employer2", OtherBasisOperation, PotentiallyCeased, BigDecimal(321.12), BigDecimal(0), BigDecimal(0))
+        )))
+
+      val mockEmploymentService = mock[EmploymentService]
+      when(mockEmploymentService.employments(any[Nino], any[TaxYear])(any[HeaderCarrier]))
+        .thenReturn(Future.successful(employments))
+
+      val sut = createSUT(incomeService = mockIncomeService, employmentService = mockEmploymentService, authentication = loggedInAuthenticationPredicate)
+      val result = sut.matchedTaxCodeIncomesForYear(nino, TaxYear().next, "EmploymentIncome", "PotentiallyCeased")(FakeRequest())
+
+      status(result) mustBe OK
+      contentAsJson(result) mustBe expectedJsonEmpty
+
+    }
+
+    "return list of live and matched pension TaxCodeIncomes for a given year" in {
+      val mockIncomeService = mock[IncomeService]
+      when(mockIncomeService.taxCodeIncomes(any(), Matchers.eq(TaxYear().next))(any()))
+        .thenReturn(Future.successful(taxCodeIncomes))
+
+      val mockEmploymentService = mock[EmploymentService]
+      when(mockEmploymentService.employments(any[Nino], any[TaxYear])(any[HeaderCarrier]))
+        .thenReturn(Future.successful(employments))
+
+      val sut = createSUT(incomeService = mockIncomeService, employmentService = mockEmploymentService, authentication = loggedInAuthenticationPredicate)
+      val result = sut.matchedTaxCodeIncomesForYear(nino, TaxYear().next, "PensionIncome", "Live")(FakeRequest())
+
+      status(result) mustBe OK
+
+      val expectedJson =
+        Json.parse(
```

Comment:
Because the logic is in the controller, you've ended up having to test the Json where I think what you want to test is the IncomeSource is correct. By moving the logic out into a service and keeping the Json stuff here, you can have one simple Json test and the more complicated cleaner/non Json tests in the service

Reply:
Are you insistent on this happening now? Potentially a future piece of refactoring work agreed.. from our pov the priority at the mo is to get mobile-paye integrated with this new contract which will also help reduce calls to tai and in turn to nps (1 less call)

Reply:
I am going to have to insist on this. Going forward we will be the team that have to maintain this and in the current state it is not something we would be willing to do. If you could pull the business logic out into its own service, test that service independently.

Reply:
Addressed